### DISCUSSION TITLE
CORS headers on exception handler responses

### DISCUSSION THREAD

**Discussion by hongymagic** [fastapi/fastapi#8027] [Questions]
**Description**

CORS middleware and exception handler responses don't seem to play well. The body of the error response cannot be read because of missing CORS headers.

Is this intentional? Given the error reposnes contain readable data, one would expect this to work with CORS, especially with the provided, standard middleware and exception handlers.

https://github.com/encode/starlette/issues/617

**Comment by hongymagic** [comment]
Closed as per encode/starlette#617, we can handle this at application tier.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue :+1:

**Comment by gs11** [comment]
Stumbled upon the same issue, how did you solve this in the app? @hongymagic 

Found an implemented a solution from here - https://github.com/tiangolo/fastapi/issues/775#issuecomment-592946834

**Comment by thomasleveil** [comment]
The solution is to wrap the FastAPI app with the Starlette [CORSMiddleware](https://www.starlette.io/middleware/#corsmiddleware) :

```python
from typing import Optional
from starlette.middleware.cors import CORSMiddleware
from fastapi import FastAPI

fastapi_app = FastAPI()
app = CORSMiddleware(app=fastapi_app, allow_origins=['*'])


@fastapi_app.get("/")
def read_root():
    return {"Hello": "World"}


@fastapi_app.get("/items/{item_id}")
def read_item(item_id: int, q: Optional[str] = None):
    return {"item_id": item_id, "q": q}

``` 

Adding the CORSMiddleware in this fashion will make all http responses, even those produced by FastApi error handler, be treated by the CORSMiddleware.

**Reply by major-mayer** [reply]
I noticed that it is necessary to run the app that is returned by CORSMiddleware with Uvicorn and use the fastapi_app for everything else.
Can somebody explain how this works?
What kind of object does the CORSMiddleware class create such that it is possible to run this as an ASGI app? 
This seems like undocumented behaviour to me, or isn't it?

**Reply by thomasleveil** [reply]
[CORSMiddleware](https://github.com/encode/starlette/blob/master/starlette/middleware/cors.py#L74) is itself an [ASGI](https://asgi.readthedocs.io/en/latest/specs/main.html) app

**Reply by major-mayer** [reply]
Ah thanks a lot, I thought it would have to inherit from some Python interface.
But it seems like the ASGI interface is pretty simple.

**Reply by ZhimaoLin** [reply]
Confirmed. Basically, wrapping the FastAPI app with the `CORSMiddleware` works. Then, launch the exposed middle ware (`app` in this example) using `uvicorn`.

Also, the `CORSMiddleware` does not have to be from `starlette.middleware.cors`. `from fastapi.middleware.cors import CORSMiddleware` works.

Here is the FastAPI version in my `Pipfile`:
```
fastapi = "0.109.0"
uvicorn = {extras = ["standard"], version = "0.27.0"}
```

**Comment by gs11** [comment]
Thanks @thomasleveil. Works as a charm!

**Comment by s-ol** [comment]
@thomasleveil's solution seems very robust and works as expected, unlike the `fastapi.middleware.CORSMiddleware` solution in the official docs.

I think that either the "official" middleware needs an update to react to the user demand, or at the very least the docs should point out this common gotcha and a solution.

**Comment by avuletica** [comment]
If you define exception handlers, e.g:

`@app.exception_handler(HTTPException)`

you will get:

`'CORSMiddleware' object has no attribute 'exception_handler'`

So I don't think this is the way :(

**Comment by gs11** [comment]
You'll need to define the exception handler on the fastapi app, not the cors one @avuletica

**Comment by avuletica** [comment]
@gs11 I'm aware of that, but if we do:

```python
fastapi_app = FastAPI()
app = CORSMiddleware(app=fastapi_app, allow_origins=['*'])
```

our app is now `CORSMiddleware` and not `FastAPI` and that means after we do this we can't change anything app-related directly, we would have to do `app.app.something` which is not pretty.

I ended up implementing a simple method:

```python
def get_app_middleware(app: FastAPI, middleware_class: Type) -> Optional[Middleware]:
    middleware_index = None
    for index, middleware in enumerate(app.user_middleware):
        if middleware.cls == middleware_class:
            middleware_index = index
    return None if middleware_index is None else app.user_middleware[middleware_index]
```

and adding CORS to exception handler:


```python
@app.exception_handler(Exception)
async def internal_exception_handler(request, exc):
    trace = traceback.format_exc()
    logger.critical(
        f"{request.url.path} {status.HTTP_500_INTERNAL_SERVER_ERROR} Internal Server Error \n {trace}."
    )
    # add anything you want in reponse content
    response = JSONResponse(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
    cors_middleware = get_app_middleware(app=request.app, middleware_class=CORSMiddleware)

    request_origin = request.headers.get("origin", "")
    if cors_middleware and "*" in cors_middleware.options["allow_origins"]:
        response.headers["Access-Control-Allow-Origin"] = "*"
    elif cors_middleware and request_origin in cors_middleware.options["allow_origins"]:
        response.headers["Access-Control-Allow-Origin"] = request_origin

    return response
```

**Comment by orfisko** [comment]
Why is this closed? To me this is simply wrong behavior in fastapi.... It should also properly handle exceptions that are being raised when the cors middleware is being added imho.

**Reply by major-mayer** [reply]
Yes I am also looking for an official solution that doesn't include wrapping the whole FastAPI app into some middleware.

**Comment by dracozombie19** [comment]
I've been seeing this behavior still with FastAPI 0.100.1.  It makes troubleshooting requests confusing (DevTools reports these as CORS errors) and more challenging in web applications.  I was able to get it working with Starlette CORSMiddleware workaround, but it seems like this would be something nice for the framework (FastAPI) to handle so that we don't need to think about this in every app.

**Reply by hamidb80** [reply]
yeah same here, it is not only confusing but also stupid

**Reply by MatheusVp2** [reply]
Yes, I had the same problem

**Comment by MatheusVp2** [comment]
In my case I'm on v0.111.0

Where do I notice that a simple get route does not register the headers

Neither by the browser's get directly on the route or by insominia or postman, nor by the request try from /docs

On my friend's PC, when he configures it, it seems to be normal, the CORS Middleware works fine.

However, when I make a request via javascript fetch in an html, the header appears

**Reply by MatheusVp2** [reply]
Direct Browser
![image](https://github.com/tiangolo/fastapi/assets/12617396/480284bb-bc71-4615-88de-c643821eebc3)
Direct Insominia
![image](https://github.com/tiangolo/fastapi/assets/12617396/e789cd92-f64e-4dee-9cd1-a6a060c2f5d9)
Direct Html Fetch Javascript
![image](https://github.com/tiangolo/fastapi/assets/12617396/5ef8d97b-a613-4d51-bd2c-d0ee9bb86e10)

**Comment by ddahan** [comment]
I'm kind of surprised this is not documented at all, I spent lots of time trying to debug this until I found this issue.
Anyway, here is my workaround:

**config.py**

```python
class Settings(BaseSettings):
    # ...
    CORS_ALLOW_ORIGIN: list[str]  # from .env file
    CORS_ALLOW_METHODS: list[str] = ["*"]
    CORS_ALLOW_HEADERS: list[str] = ["*"]
    CORS_ALLOW_CREDENTIALS: bool = True
```

**main.py**

```python
app.add_middleware(
    CORSMiddleware,
    allow_origins=settings.CORS_ALLOW_ORIGIN,
    allow_credentials=settings.CORS_ALLOW_CREDENTIALS,
    allow_methods=settings.CORS_ALLOW_METHODS,
    allow_headers=settings.CORS_ALLOW_HEADERS,
)
```

**exceptions.py**

```python
def add_cors(request: Request, response: JSONResponse) -> JSONResponse:
    """
    CORS headers are not automatically added to error handlers with FastAPI
    cf. https://github.com/fastapi/fastapi/discussions/8027
    """
    cors_headers = {
        "Access-Control-Allow-Methods": ", ".join(settings.CORS_ALLOW_METHODS),
        "Access-Control-Allow-Headers": ", ".join(settings.CORS_ALLOW_HEADERS),
        "Access-Control-Allow-Credentials": str(settings.CORS_ALLOW_CREDENTIALS).lower(),
    }

    # If the origin is in the allowed list, add it to the CORS headers
    origin = request.headers.get("origin")
    if origin in settings.CORS_ALLOW_ORIGIN:
        cors_headers["Access-Control-Allow-Origin"] = origin

    response.headers.update(cors_headers)
    return response

def internal_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    response = JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content={
            "errors": {
                "general": [
                    "An unknown error has occurred. Please contact an administrator."
                ]
            }
        },
    )
    return add_cors(request, response)
```

**Comment by timothy-jeong** [comment]
This issue occurs due to Starlette's middleware chain sequence. Please refer to [Starlette #2876](https://github.com/encode/starlette/discussions/2876) for more details. The official solution is to wrap the application with CORSMiddleware, as documented [here](https://www.starlette.io/middleware/#corsmiddleware-global-enforcement).

**Comment by KSmanis** [comment]
An alternative approach to wrapping the FastAPI app without changing its type to `CORSMiddleware`:
```python
from fastapi import FastAPI as FastAPIBase


class FastAPI(FastAPIBase):
    def build_middleware_stack(self) -> ASGIApp:
        return CORSMiddleware(super().build_middleware_stack(), allow_origins=["*"])


app = FastAPI()
```
Works as-is with Starlette apps as well.

**Comment by DINGdeZHENZ** [comment]
Just a example for copying in future
```
from fastapi.responses import JSONResponse
from starlette.middleware.cors import CORSMiddleware
from fastapi import FastAPI, Request
import uvicorn

app = FastAPI()
appWithCORS = CORSMiddleware(app=app, allow_origins=['*'])


@app.get("/")
def read_root():
    return {"Hello": "World"}
@app.get("/error")
def error():
    raise RuntimeError("call me error")
@app.exception_handler(Exception)
async def all_exception_handler(request: Request, exc: Exception):
     return JSONResponse(
            status_code=500,
            content={
                "code": 500,
                "message": str(exc),
            }
        )

if __name__ == "__main__":
    uvicorn.run(appWithCORS, port=80)
```
